diff -Naur a/gettext-tools/gnulib-lib/basename.h b/gettext-tools/gnulib-lib/basename.h
--- a/gettext-tools/gnulib-lib/basename.h	2016-03-20 10:37:53.000000000 +0300
+++ b/gettext-tools/gnulib-lib/basename.h	2025-03-12 16:30:43.455168884 +0300
@@ -20,6 +20,7 @@
 
 /* This is where basename() is declared.  */
 #include <string.h>
+#include <libgen.h>
 
 
 #ifdef __cplusplus
diff -Naur a/gettext-runtime/gnulib-lib/basename.h b/gettext-runtime/gnulib-lib/basename.h
--- a/gettext-runtime/gnulib-lib/basename.h	2016-03-20 10:37:53.000000000 +0300
+++ b/gettext-runtime/gnulib-lib/basename.h	2025-03-12 16:30:43.455168884 +0300
@@ -20,6 +20,7 @@
 
 /* This is where basename() is declared.  */
 #include <string.h>
+#include <libgen.h>
 
 
 #ifdef __cplusplus
diff -Naur a/gettext-tools/libgettextpo/basename.h b/gettext-tools/libgettextpo/basename.h
--- a/gettext-tools/libgettextpo/basename.h	2016-03-20 10:37:53.000000000 +0300
+++ b/gettext-tools/libgettextpo/basename.h	2025-03-12 16:30:43.455168884 +0300
@@ -20,6 +20,7 @@
 
 /* This is where basename() is declared.  */
 #include <string.h>
+#include <libgen.h>
 
 
 #ifdef __cplusplus
diff -Naur a/gettext-tools/src/locating-rule.c b/gettext-tools/src/locating-rule.c
--- a/gettext-tools/src/locating-rule.c	2016-03-20 10:37:53.000000000 +0300
+++ b/gettext-tools/src/locating-rule.c	2025-03-12 16:15:34.691399024 +0300
@@ -27,6 +27,8 @@
 #include "concat-filename.h"
 #include "c-strcase.h"
 
+#include <stdlib.h>
+
 #if HAVE_DIRENT_H
 # include <dirent.h>
 #endif
