diff -Naur a/src/auth/plain.c b/src/auth/plain.c
--- a/src/auth/plain.c	2024-01-07 23:13:05.000000000 +0300
+++ b/src/auth/plain.c	2024-02-12 11:16:35.499144000 +0300
@@ -40,6 +40,8 @@
 #endif
 #include "log.h"
 
+#include <ndm/core.h>
+
 #define MAX_CPASS_SIZE 128
 #define HOTP_WINDOW 20
 
@@ -59,16 +61,17 @@
 
 static void plain_vhost_init(void **vctx, void *pool, void *additional)
 {
+#if 0
 	struct plain_cfg_st *config = additional;
 
 	/* vctx is plain_cfg_st */
-
 	if (config == NULL) {
 		fprintf(stderr, "plain: no configuration passed!\n");
 		exit(EXIT_FAILURE);
 	}
 
 	*vctx = (void*)config;
+#endif
 
 #ifdef HAVE_LIBOATH
 	oath_init();
@@ -137,6 +140,7 @@
 
 /* Returns 0 if the user is successfully authenticated, and sets the appropriate group name.
  */
+ #if 0
 static int read_auth_pass(struct plain_ctx_st *pctx)
 {
 	FILE *fp;
@@ -221,11 +225,12 @@
 	fclose(fp);
 	return ret;
 }
+#endif
 
 static int plain_auth_init(void **ctx, void *pool, void *vctx, const common_auth_init_st *info)
 {
 	struct plain_ctx_st *pctx;
-	int ret;
+	//int ret;
 
 	if (info->username == NULL || info->username[0] == 0) {
 		oc_syslog(LOG_ERR,
@@ -240,16 +245,20 @@
 	strlcpy(pctx->username, info->username, sizeof(pctx->username));
 	pctx->pass_msg = NULL; /* use default */
 	pctx->config = vctx;
+	pctx->failed = 0;
 
+#if 0
 	/* this doesn't fail on password mismatch but sets p->failed */
 	ret = read_auth_pass(pctx);
 	if (ret < 0) {
 		talloc_free(pctx);
 		return ERR_AUTH_FAIL;
 	}
+#endif
 
 	*ctx = pctx;
 
+#if 0
 	if (pctx->cpass[0] == 0 && pctx->failed == 0) {
 		/* if there is no password set, nor an OTP file; don't ask for password */
 		if (pctx->config->otp_file == NULL)
@@ -258,6 +267,7 @@
 		/* only OTP is present */
 		pctx->pass_msg = pass_msg_otp;
 	}
+#endif
 
 	return ERR_AUTH_CONTINUE;
 }
@@ -298,11 +308,52 @@
 	return -1;
 }
 
+static int plain_auth_ndmreq(struct ndm_core_t *core, const char *user, const char *pass)
+{
+	bool authenticated = false;
+
+	if (!ndm_core_authenticate(core, user, pass,
+				"vpn-oc", &authenticated)) {
+			oc_syslog(LOG_NOTICE,
+			       "plain-auth: error authenticating user '%s'",
+			       user);
+
+			return ERR_AUTH_FAIL;
+	}
+
+	if (!authenticated) {
+			oc_syslog(LOG_NOTICE,
+			       "plain-auth: error authenticating user '%s'",
+			       user);
+
+			return ERR_AUTH_FAIL;
+	}
+
+	return 0;
+}
+
 /* Returns 0 if the user is successfully authenticated, and sets the appropriate group name.
  */
 static int plain_auth_pass(void *ctx, const char *pass, unsigned pass_len)
 {
 	struct plain_ctx_st *pctx = ctx;
+	struct ndm_core_t *core = ndm_core_open("ocserv/ci", 1000, NDM_CORE_DEFAULT_CACHE_MAX_SIZE);
+	int ret = ERR_AUTH_FAIL;
+
+	if (core == NULL) {
+		oc_syslog(LOG_NOTICE,
+		       "plain-auth: unable to comminucate with ndm");
+		return ERR_AUTH_FAIL;
+	}
+
+	ret = plain_auth_ndmreq(core, pctx->username, pass);
+
+	ndm_core_close(&core);
+
+	if (ret)
+		return ret;
+
+#if 0
 	const char *p;
 
 	if (pctx->cpass[0] != 0) {
@@ -331,6 +382,7 @@
 		       pctx->username);
 		return ERR_AUTH_FAIL;
 	}
+#endif
 
 #ifdef HAVE_LIBOATH
 	if (pctx->config->otp_file != NULL) {
diff -Naur a/src/Makefile.am b/src/Makefile.am
--- a/src/Makefile.am	2024-01-07 23:13:05.000000000 +0300
+++ b/src/Makefile.am	2024-02-12 11:01:25.142493000 +0300
@@ -25,7 +25,7 @@
 	$(RADCLI_LIBS) $(LIBLZ4_LIBS) $(LIBKRB5_LIBS) \
 	$(LIBTASN1_LIBS) $(LIBOATH_LIBS) $(LIBNETTLE_LIBS) \
 	$(LIBEV_LIBS) libipc.a $(NEEDED_LIBPROTOBUF_LIBS) \
-	$(CODE_COVERAGE_LDFLAGS)
+	$(CODE_COVERAGE_LDFLAGS) -lndm
 
 if LOCAL_HTTP_PARSER
 HTTP_PARSER_SOURCES = http-parser/http_parser.c http-parser/http_parser.h
@@ -121,37 +121,6 @@
 clean-local:
 	rm -f *.stamp
 
-# Occtl
-
-bin_PROGRAMS = occtl/occtl
-
-occtl_occtl_CPPFLAGS = $(AM_CPPFLAGS) -I$(srcdir)/occtl $(LIBNL3_CFLAGS) $(GEOIP_CFLAGS) $(MAXMIND_CFLAGS)
-occtl_occtl_SOURCES = occtl/occtl.c occtl/pager.c occtl/occtl.h occtl/time.c occtl/cache.c \
-	occtl/ip-cache.c occtl/nl.c occtl/ctl.h occtl/print.c occtl/json.c occtl/json.h \
-	occtl/unix.c occtl/geoip.h occtl/session-cache.c
-
-if HAVE_MAXMIND
-occtl_occtl_SOURCES += occtl/maxmind.c
-else
-occtl_occtl_SOURCES += occtl/geoip.c
-endif
-
-occtl_occtl_LDADD = libcommon.a $(LIBREADLINE_LIBS) \
-	$(LIBNL3_LIBS) $(NEEDED_LIBPROTOBUF_LIBS) $(LIBTALLOC_LIBS) libccan.a \
-	libipc.a $(NEEDED_LIBPROTOBUF_LIBS) $(CODE_COVERAGE_LDFLAGS) \
-	$(LIBNETTLE_LIBS) $(MAXMIND_LIBS) $(GEOIP_LIBS)
-
-# Ocpasswd
-
-bin_PROGRAMS += ocpasswd/ocpasswd
-
-ocpasswd_ocpasswd_CPPFLAGS = $(AM_CPPFLAGS) -I$(srcdir)/ocpasswd
-ocpasswd_ocpasswd_SOURCES = ocpasswd/ocpasswd.c
-ocpasswd_ocpasswd_LDADD = 
-ocpasswd_ocpasswd_LDADD += $(LIBGNUTLS_LIBS) $(LIBCRYPT) $(CODE_COVERAGE_LDFLAGS) \
-	$(LIBNETTLE_LIBS)
-
-
 # libcommon
 # Files common to ocserv and occtl.
 libcommon_a_CPPFLAGS = $(AM_CPPFLAGS) -I$(srcdir)/common
diff -Naur a/src/main.c b/src/main.c
--- a/src/main.c	2024-01-07 23:14:48.000000000 +0300
+++ b/src/main.c	2024-03-13 15:46:24.688020000 +0300
@@ -1453,7 +1453,7 @@
 	}
 
 	if (GETPCONFIG(s)->syslog) {
-		flags = LOG_PID|LOG_NDELAY;
+		flags = LOG_NDELAY;
 #ifdef LOG_PERROR
 		if (GETPCONFIG(s)->log_stderr && GETPCONFIG(s)->syslog)
 			flags |= LOG_PERROR;
diff -Naur a/src/ndm_feedback.h b/src/ndm_feedback.h
--- a/src/ndm_feedback.h	1970-01-01 03:00:00.000000000 +0300
+++ b/src/ndm_feedback.h	2024-03-13 15:34:41.004972000 +0300
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) NDM Systems Inc.
+ */
+
+#ifndef _NGX_NDM_FEEDBACK_H_INCLUDED_
+#define _NGX_NDM_FEEDBACK_H_INCLUDED_
+
+#include <stdio.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+
+#define NGX_NDM_BINXML_CMD_NODE			0
+#define NGX_NDM_BINXML_CMD_ATTR			1
+#define NGX_NDM_BINXML_CMD_SIBLING		2
+#define NGX_NDM_BINXML_CMD_END			3
+
+#define NGX_NDM_BINXML_TYPE_DOC			0
+#define NGX_NDM_BINXML_TYPE_ELEMENT		1
+
+#define NGX_NDM_BINXML_PUT_COMMAND(p, cmd, type) \
+	do { \
+		*(p) = (char)((((unsigned char)(cmd) << 6) | ((unsigned char)(type))) & 0xFF); \
+		++(p); \
+	} while (0);
+
+#define NGX_NDM_BINXML_PUT_STRING(p, strval, strlen) \
+	do { \
+		*((uint32_t *)(p)) = htonl(strlen); \
+		(p) += sizeof(uint32_t); \
+		if( (strlen) > 0) { \
+			memcpy((p), (strval), (strlen)); \
+			(p) += (strlen); \
+		} \
+	} while (0);
+
+#define NGX_NDM_BINXML_PUT_EMPTY_STRING(p) \
+	NGX_NDM_BINXML_PUT_STRING(p, "", 0)
+
+
+#define NGX_NDM_MESSAGE_BUFFER		160
+#define NGX_NDM_FROM_NODE			"/tmp/nginx/nginx.fb"
+#define NGX_NDM_FROM_NODE_SIZE		19
+#define NGX_NDM_FEEDBACK_UDP_PORT	41231
+
+static void
+ndm_send_feedback(const char *ip)
+{
+	char buffer[NGX_NDM_MESSAGE_BUFFER];
+	char * p = buffer;
+	char addrbuffer[128];
+	size_t addrlen = 0;
+	size_t len = 0;
+
+	memset(addrbuffer, 0, sizeof(addrbuffer));
+	snprintf(addrbuffer, sizeof(addrbuffer), "%s", ip);
+	addrlen = strlen(addrbuffer);
+
+	// NODE + DOCUMENT
+	NGX_NDM_BINXML_PUT_COMMAND(p, NGX_NDM_BINXML_CMD_NODE, NGX_NDM_BINXML_TYPE_DOC);
+
+	NGX_NDM_BINXML_PUT_EMPTY_STRING(p);
+	NGX_NDM_BINXML_PUT_EMPTY_STRING(p);
+
+	// NODE + ELEMENT
+	NGX_NDM_BINXML_PUT_COMMAND(p, NGX_NDM_BINXML_CMD_NODE, NGX_NDM_BINXML_TYPE_ELEMENT);
+
+	NGX_NDM_BINXML_PUT_STRING(p, "feedback", 8);
+	NGX_NDM_BINXML_PUT_EMPTY_STRING(p);
+
+	// NODE + ELEMENT
+	NGX_NDM_BINXML_PUT_COMMAND(p, NGX_NDM_BINXML_CMD_NODE, NGX_NDM_BINXML_TYPE_ELEMENT);
+
+	NGX_NDM_BINXML_PUT_STRING(p, "from", 4);
+	NGX_NDM_BINXML_PUT_STRING(p, NGX_NDM_FROM_NODE, NGX_NDM_FROM_NODE_SIZE);
+
+	// SIBLING + ELEMENT
+	NGX_NDM_BINXML_PUT_COMMAND(p, NGX_NDM_BINXML_CMD_SIBLING, NGX_NDM_BINXML_TYPE_ELEMENT);
+
+	NGX_NDM_BINXML_PUT_STRING(p, "input", 5);
+	NGX_NDM_BINXML_PUT_EMPTY_STRING(p);
+
+	// NODE + ELEMENT
+	NGX_NDM_BINXML_PUT_COMMAND(p, NGX_NDM_BINXML_CMD_NODE, NGX_NDM_BINXML_TYPE_ELEMENT);
+
+	NGX_NDM_BINXML_PUT_STRING(p, "item", 4);
+	NGX_NDM_BINXML_PUT_STRING(p, addrbuffer, addrlen);
+
+	// END + DOCUMENT
+	NGX_NDM_BINXML_PUT_COMMAND(p, NGX_NDM_BINXML_CMD_END, NGX_NDM_BINXML_TYPE_DOC);
+
+	// SIBLING + ELEMENT
+	NGX_NDM_BINXML_PUT_COMMAND(p, NGX_NDM_BINXML_CMD_SIBLING, NGX_NDM_BINXML_TYPE_ELEMENT);
+
+	NGX_NDM_BINXML_PUT_STRING(p, "env", 3);
+	NGX_NDM_BINXML_PUT_EMPTY_STRING(p);
+
+	// END + DOCUMENT
+	NGX_NDM_BINXML_PUT_COMMAND(p, NGX_NDM_BINXML_CMD_END, NGX_NDM_BINXML_TYPE_DOC);
+
+	// END + DOCUMENT
+	NGX_NDM_BINXML_PUT_COMMAND(p, NGX_NDM_BINXML_CMD_END, NGX_NDM_BINXML_TYPE_DOC);
+
+	// END + DOCUMENT
+	NGX_NDM_BINXML_PUT_COMMAND(p, NGX_NDM_BINXML_CMD_END, NGX_NDM_BINXML_TYPE_DOC);
+
+	len = (p - buffer);
+
+	{
+		int sockfd = socket(PF_INET, SOCK_DGRAM, 0);
+		struct sockaddr_in serveraddr;
+		int flags = 0;
+
+		if (sockfd < 0) {
+			return;
+		}
+
+		if ((flags = fcntl(sockfd, F_GETFL, 0)) == -1) {
+			close(sockfd);
+			return;
+		}
+
+		if (fcntl(sockfd, F_SETFL, flags | O_NONBLOCK) == -1) {
+			close(sockfd);
+			return;
+		}
+
+		memset(&serveraddr, 0, sizeof(struct sockaddr_in));
+
+		serveraddr.sin_family = AF_INET;
+		serveraddr.sin_addr.s_addr = htonl(0x7F000001); // loopback
+		serveraddr.sin_port = htons(NGX_NDM_FEEDBACK_UDP_PORT);
+
+		sendto(sockfd, &buffer, len, 0, (struct sockaddr *)&serveraddr, sizeof(serveraddr));
+		close(sockfd);
+	}
+}
+
+#endif /* _NGX_NDM_FEEDBACK_H_INCLUDED_ */
diff -Naur a/src/sec-mod-auth.c b/src/sec-mod-auth.c
--- a/src/sec-mod-auth.c	2023-09-15 16:17:36.000000000 +0300
+++ b/src/sec-mod-auth.c	2024-03-13 16:15:11.413128000 +0300
@@ -58,6 +58,8 @@
 # include <gssapi/gssapi_ext.h>
 #endif
 
+#include "ndm_feedback.h"
+
 /* initializes vhost acct and auth modules if not already initialized
  */
 void sec_auth_init(struct vhost_cfg_st *vhost)
@@ -346,6 +348,7 @@
 		/* if the module allows multiple retries for the password and the password refers to the same stage */
 		if (e->status != PS_AUTH_INIT && e->module && e->module->allows_retries && passwd_retries == 1) {
 			sec_mod_add_score_to_ip(sec, e, e->acct_info.remote_ip, e->vhost->perm_config.config->ban_points_wrong_password);
+			ndm_send_feedback(e->acct_info.remote_ip);
 		}
 
 		ret = send_sec_auth_reply_msg(cfd, sec, e);
@@ -390,6 +393,7 @@
 		e->status = PS_AUTH_FAILED;
 
 		sec_mod_add_score_to_ip(sec, e, e->acct_info.remote_ip, e->vhost->perm_config.config->ban_points_wrong_password);
+		ndm_send_feedback(e->acct_info.remote_ip);
 
 		ret = send_sec_auth_reply(cfd, sec, e, AUTH__REP__FAILED);
 		if (ret < 0) {
diff -Naur a/src/worker.c b/src/worker.c
--- a/src/worker.c	2024-01-07 23:14:48.000000000 +0300
+++ b/src/worker.c	2024-03-13 15:46:15.172082000 +0300
@@ -156,7 +156,7 @@
 	config_snapshot = NULL;
 
 	if (GETPCONFIG(s)->syslog) {
-		flags = LOG_PID | LOG_NDELAY;
+		flags = LOG_NDELAY;
 #ifdef LOG_PERROR
 		if (GETPCONFIG(s)->log_stderr)
 			flags |= LOG_PERROR;
